partiql-logical-planner/src/typer.rs

```diff
@@ -0,0 +1,917 @@
+        let err2 = r#"Unexpected Typing Environment; expected typing environment with only one type but found 2 types"#;
+
+        assert_err(
+            assert_query_typing(
+                TypingMode::Strict,
+                "SELECT customers.id, customers.name, customers.age FROM customers",
+                create_customer_schema(
+                    false,
+                    vec![
+                        StructField::new("id", int!()),
+                        StructField::new("name", str!()),
+                    ],
+                ),
+                vec![],
+            ),
+            vec![
+                TypingError::TypeCheck(err1.to_string()),
+                TypingError::IllegalState(err2.to_string()),
+            ],
+            Some(bag![r#struct![BTreeSet::from([StructConstraint::Fields(
+                vec![
+                    StructField::new("id", int!()),
+                    StructField::new("name", str!()),
+                    StructField::new("age", undefined!()),
+                ]
+            ),])]]),
+        );
+
+        // Closed Schema with `Strict` typing mode and `bar` non-existent projection from closed nested `details`.
+        let details_fields = struct_fields![("age", int!())];
+        let details = r#struct![BTreeSet::from([
+            details_fields,
+            StructConstraint::Open(false)
+        ])];
+
+        let err1 = r#"No Typing Information for SymbolPrimitive { value: "details", case: CaseInsensitive } in closed Schema PartiqlType(Struct(StructType { constraints: {Open(false), Fields([StructField { name: "age", ty: PartiqlType(Int) }])} }))"#;
+        let err2 = r#"No Typing Information for SymbolPrimitive { value: "bar", case: CaseInsensitive } in closed Schema PartiqlType(Struct(StructType { constraints: {Open(false), Fields([StructField { name: "age", ty: PartiqlType(Int) }])} }))"#;
+
+        assert_err(
+            assert_query_typing(
+                TypingMode::Strict,
+                "SELECT customers.id, customers.name, customers.details.bar FROM customers",
+                create_customer_schema(
+                    false,
+                    vec![
+                        StructField::new("id", int!()),
+                        StructField::new("name", str!()),
+                        StructField::new("details", details),
+                    ],
+                ),
+                vec![],
+            ),
+            vec![
+                TypingError::TypeCheck(err1.to_string()),
+                TypingError::TypeCheck(err2.to_string()),
+            ],
+            Some(bag![r#struct![BTreeSet::from([StructConstraint::Fields(
+                vec![
+                    StructField::new("id", int!()),
+                    StructField::new("name", str!()),
+                    StructField::new("bar", int!()),
```

Comment:
How is `bar` resolved to `int` here? Shouldn't this be `missing` since `customers.details` only has field `age`?

Reply:
This is in error state, so the schema is expected to be incorrect, however the more correct erroneous type for this is `UNDEFINED` which can translate to `MISSING` in permissive mode. Let me look further into this.